services/ExtraActivityManager.php

```diff
@@ -0,0 +1,352 @@
+        return $this->getExtraActivityLogsFromLike($like, (!is_null($learner))?$learner->getUsername():'') ;
+    }
+
+    /**
+     * Get a Extra-activity from tag
+     * @param string $tag tag of the ExtraActivityLog
+     * @return ExtraActivityLog|null return null if empty tag
+     */
+    public function getExtraActivityLog(string $tag): ?ExtraActivityLog
+    {
+        if (empty($tag)) {
+            return null;
+        }
+        $like = '%"tag":"' . $tag . '"%';
+        $extraActivities = $this->getExtraActivityLogsFromLike($like) ;
+        return $extraActivities->get($tag) ;
+    }
+
+    /**
+     * Get extra-activitylogs from like query
+     * @param string $like query to send to TripleStore->getMatching
+     * @param string $learnerName
+     * @return ExtraActivityLogs the courseStructure's extraActivities
+     */
+    private function getExtraActivityLogsFromLike(string $like, string $learnerName = ''):ExtraActivityLogs
+    {
+        $results = $this->tripleStore->getMatching(
+            (empty($learnerName) ? null : $learnerName),
+            self::LMS_TRIPLE_PROPERTY_NAME_EXTRA_ACTIVITY,
+            $like,
+            (empty($learnerName) ? 'LIKE' : '='),
+            '=',
+            'LIKE'
+        );
+
+        $extraActivities = new ExtraActivityLogs([]) ;
+        if ($results) {
+            foreach ($results as $result) {
+                $extraActivity = ExtraActivityLog::createFromJSON(
+                    $result['value'],
+                    $this->courseManager,
+                    $this->dateManager
+                );
+                if ($extraActivity) {
+                    if (!$extraActivities->add($extraActivity)) {
+                        // already present
+                        $extraActivity = $extraActivities->get($extraActivity->getTag());
+                    };
+                    $extraActivity->addLearnerName($result['resource']) ;
+                }
+            }
+        }
+        return $extraActivities ;
+    }
+
+    
+    /**
+     * delete the Extra-activities of a courseStructure
+     * @param string $tag tag of the ExtraActivityLog
+     * @param string $learnerName (optional), if not present, delete for all learners
+     * @return bool false in cas of error
+     */
+    public function deleteExtraActivity(string $tag, string $learnerName = ''): bool
+    {
+        $like = '%"tag":"' . $tag . '"%';
+        $results = $this->tripleStore->getMatching(
+            (empty($learnerName) ? null : $learnerName),
+            self::LMS_TRIPLE_PROPERTY_NAME_EXTRA_ACTIVITY,
+            $like,
+            (empty($learnerName) ? 'LIKE' : '='),
+            '=',
+            'LIKE'
+        );
+
+        if (!$results) {
+            if ($debug) {
+                throw new \Exception('Errors in '. get_class($this) . ' : not possible to delete tag : "'.$tag.'" because not existing', 1);
+            }
+            return false ;
+        }
+
+        $log = true ;
```

Comment:
enlever la variable $log, mettre return false après $debug, et retourner true à la fin de fonction -> plus clair et permet d'envoyer l'erreur direct sans avoir à itérer sur les éléments suivants